### PR TITLE
docs: HACKING.md: add custom payload build section

### PR DESCRIPTION
**Pull request by runcom**
Please go over it by testing the flow as well to make sure it's correct (I just did and it works for me (tm)).
Please also correct any of my mistakes in english lol

Signed-off-by: Antonio Murdaca <runcom@linux.com>

### CODE REVIEW

File: docs/HACKING.md [hunk 1]
```diff
@@ -200,3 +200,54 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
 
 And from there debug the cluster live while the tests are running. Though be
 aware that it will be quickly torn down as soon as the tests succeed or fail.
+
+# Building and installing a custom release image for tesing the MCO
```
Comment (cgwalters) [inline docs/HACKING.md]: Thanks for adding this!

File: docs/HACKING.md [hunk 2]
```diff
@@ -200,3 +200,54 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+1) build the MCO components images you're interested into testing
+2) build a custom release payload
+3) install a cluster with your custom release payload
+
+## Build the MCO components images
+
+To build the image for any of the component run:
+
+```
+podman build -t quay.io/user/machine-config-{component} -f Dockerfile.machine-config-{component} .
```
Comment (cgwalters) [inline docs/HACKING.md]: There's also `make image-controller` `make image-daemon` etc.

Comment (runcom) [reply docs/HACKING.md]: omg, how have I missed those :disappointed_relieved:

File: docs/HACKING.md [hunk 3]
```diff
@@ -200,3 +200,58 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
 
 And from there debug the cluster live while the tests are running. Though be
 aware that it will be quickly torn down as soon as the tests succeed or fail.
+
+# Building and installing a custom release image for tesing the MCO
```
Comment (kikisdeliveryservice) [inline docs/HACKING.md]: "tesing" -> "testing"

File: docs/HACKING.md [hunk 4]
```diff
@@ -200,3 +200,58 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+
+During the MCO development there are certain test scenarios for any of the component
```
Comment (kikisdeliveryservice) [inline docs/HACKING.md]: i think delete "for any of the component"

File: docs/HACKING.md [hunk 5]
```diff
@@ -200,3 +200,58 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+that would require a cluster running with a custom release to properly test your
```
Comment (kikisdeliveryservice) [inline docs/HACKING.md]: "that would" -> "that"

File: docs/HACKING.md [hunk 6]
```diff
@@ -200,3 +200,58 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+code (i.e. test the interaction of the MCO with the CVO).
+
+In those situations, you will need to:
+
+1) build the MCO components images you're interested into testing
```
Comment (kikisdeliveryservice) [inline docs/HACKING.md]: "into" -> "in"

File: docs/HACKING.md [hunk 7]
```diff
@@ -200,3 +200,58 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+2) build a custom release payload
+3) install a cluster with your custom release payload
+
+## Build the MCO components images
+
+To build the image for any of the component run:
```
Comment (kikisdeliveryservice) [inline docs/HACKING.md]: "any of the" -> "any"

**Comment by kikisdeliveryservice** [comment]
Generally, I'm wondering at what point we start breaking some of these sections out into their own pages and just link to them in HACKING.MD bc it's getting long.

**Comment by runcom** [comment]
> Generally, I'm wondering at what point we start breaking some of these sections out into their own pages and just link to them in HACKING.MD bc it's getting long.

I can do that as part of this PR if you want to, any idea on how to break it down to different pages?

**Comment by kikisdeliveryservice** [comment]
@runcom I'm ok with not doing it here. I think once this is in, we can take stock of what we have and decide how we'd like to organize it going forward. :smile:

File: docs/HACKING.md [hunk 8]
```diff
@@ -200,3 +200,59 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+
+## Build a custom release payload
+
+Now that your have your custom component images, to build a custom release payload, run:
+
+```
+oc adm release new -n openshift --server https://api.ci.openshift.org \
+                                --from-image-stream "origin-v4.0" \
+                                --to-image quay.io/use/origin-release:v4.0 \
+                                machine-config-{component}=quay.io/user/machine-config-{component}
```
Comment (runcom) [inline docs/HACKING.md]: there's a known limitation now in code for this though https://github.com/openshift/machine-config-operator/issues/421, which can be worked around with https://github.com/openshift/machine-config-operator/issues/421 (basically all operands have to be at the very same version or the installation just fails).
That can be worked around by bundling all the components though (see https://github.com/openshift/machine-config-operator/issues/421#issuecomment-463392262)

**Comment by runcom** [comment]
/hold

Holding due to https://github.com/openshift/machine-config-operator/pull/425/files#r256622962 but please keep testing this :tada:

File: docs/HACKING.md [hunk 9]
```diff
@@ -200,3 +200,59 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+After the build is complete, make sure to push the image to a registry (i.e. `quay.io/user/machine-config-{component}`).
+
+## Build a custom release payload
+
+Now that your have your custom component images, to build a custom release payload, run:
+
+```
+oc adm release new -n openshift --server https://api.ci.openshift.org \
+                                --from-image-stream "origin-v4.0" \
+                                --to-image quay.io/use/origin-release:v4.0 \
```
Comment (jlebon) [inline docs/HACKING.md]: `user`

File: docs/HACKING.md [hunk 10]
```diff
@@ -200,3 +200,59 @@ export KUBECONFIG=$XDG_RUNTIME_DIR/kubeconfig
...
+When the command above finishes, your custom release payload is going to be available
+at the location you specified via the `--to-image` flag for the installer to be consumed.
+
+Note for quay.io users: images pushed to your personal account are going to be
+private by default. If you want to keep the release payload image private you will need to provide
+the secret to pull it in the `install-config.yaml` configuration (`pullSecret` section specifically).
+
+## Install a cluster with a custom release payload
+
+In order to use your new custom release payload to install a new cluster, simple run the creation process with
```
Comment (jlebon) [inline docs/HACKING.md]: simply

File: docs/HACKING.md (hunk 8 above)
Comment (jlebon) [reply docs/HACKING.md]: Probably worth just mentioning this limitation and link to the issue directly in the docs.

**Comment by runcom** [comment]
/hold cancel

because I've addressed https://github.com/openshift/machine-config-operator/pull/425#discussion_r256889135

File: docs/HACKING.md (hunk 8 above)
Comment (runcom) [reply docs/HACKING.md]: addressed

**Comment by kikisdeliveryservice** [comment]
I'm a little confused here what am I supposed to do for this particular part
` --to-image quay.io/user/origin-release:v4.0 \` 

I get building the mco components and adding but am I supposed to somehow build origin-release:v4.0??

**Comment by runcom** [comment]
> I get building the mco components and adding but am I supposed to somehow build origin-release:v4.0??

that image is the output of `oc adm release new` itself, it gets produced as part of that command and gets automagically pushed as well so you can use it with the installer

**Comment by runcom** [comment]
repushed cause I've missed some backslashes (thanks @kikisdeliveryservice )

**Comment by rphillips** [comment]
Thanks for writing this up. Looks awesome!

**Pull request merged**